src/models/tabular_flows.jl

```diff
@@ -0,0 +1,115 @@
+using Flux
+using StatsBase
+using Distributions
+using DistributionsAD
+using ContinuousFlows: RealNVP, MaskedAutoregressiveFlow
+using ValueHistories
+using MLDataPattern: RandomBatches
+
+abstract type TabularFlow end
+
+struct RealNVPFlow <: TabularFlow
+	flows
+	base
+end
+
+function RealNVPFlow(nflows::Int, isize::Int, hsize::Int, nlayers::Int)
+	RealNVPFlow(Chain([
+		RealNVP(
+			isize, 
+			(d, o, ftype, postprocess) -> build_mlp(d, hsize, o, nlayers, ftype=ftype, lastlayer=postprocess),
+			mod(i,2) == 0)
+		for i in 1:nflows]...), MvNormal(isize, 1.0f0))
+end
+
+(nvpf::RealNVPFlow)(X) = nvpf.flows(X)
+Flux.trainable(nvpf::RealNVPFlow) = (nvpf.flows, )
+
+struct MAF <: TabularFlow
+	flows
+	base
+end
+
+function MAF(nflows::Int, isize::Int, hsize::Int, nlayers::Int, ordering::String)
+	MAF(Chain([
+        MaskedAutoregressiveFlow(
+            isize, 
+            hsize,
+            nlayers, 
+            isize, 
+            (ordering == "natural") ? (
+                (mod(i, 2) == 0) ? "reversed" : "sequential"
+              ) : "random"
+            ) 
+        for i in 1:nflows]...), MvNormal(isize, 1.0f0))
+end
+
+(maf::MAF)(X) = maf.flows(X)
+Flux.trainable(maf::MAF) = (maf.flows, )
+
+
+function loss(model::F, X) where {F <: TabularFlow}
+    Z, logJ = model((X, _init_logJ(X)))
+    -sum(logpdf(model.base, Z)' .+ logJ)/size(X, 2)
+end
+
+function StatsBase.fit!(model::F, data::Tuple, p) where F <: TabularFlow
+	opt = Flux.ADAM(p.lr)
+	
+	trn_model = deepcopy(model)
+	ps = Flux.params(trn_model);
+
+	X = data[1][1]
+	X_val = data[2][1]
+
+	history = MVHistory()
+	patience = p.patience
+	reg = (p.wreg > 0) ? l2_reg : _ -> 0.0
+	
+	best_val_loss = loss(trn_model, X_val)
```

Comment:
Have you considered computing the loss over a whole validation dataset batch-wise? This might be very expensive for large datasets... or not? I definitely do it on image data where only a batch of samples fits on the gpu.

Reply:
What I mean is computing it batch-wise and then taking a mean.

Reply:
I have tested it, nad for elbo in a VAE model this starts to make sense from 5.000 samples (input dim = 4)

Reply:
I will try to measure it with some fixed model architecture. I expect the time spent there is not insignificant in comparison to the whole training loop.